File: Examples/intel_examples/examples_core_c/datafittingc/source/dfdnaturalcubicspline.c
/*******************************************************************************
* Copyright 2010-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
!  Content:
!    Construction and integration of natural cubic spline
!  Example Program Text
!******************************************************************************/

#include <stdio.h>

#include "mkl.h"
#include "errcheck.inc"
#include "generatedata.inc"
#include "rescheck.inc"

#define N              6 // number of break points
#define NY             1 // number of functions
#define NLIM           4 // number of pairs of integration limits

#define NSCOEFF        (NY*(N-1)*DF_PP_CUBIC)  // total number of spline
                                               // coefficients

#define LLIM_X       0.0 // left limit of interpolation interval
#define RLIM_X       7.0 // right limit of interpolation interval
#define LEFT_LLIM    1.0
#define RIGHT_LLIM   3.0
#define LEFT_RLIM    4.0
#define RIGHT_RLIM   5.5
#define FREQ        0.15


int main()
{
    DFTaskPtr task;                     // Data Fitting task descriptor
    MKL_INT sorder;                     // spline order
    MKL_INT stype;                      // spline type
    MKL_INT nx;                         // number of break points
    MKL_INT xhint;                      // additional info about break points
    MKL_INT ny;                         // number of functions
    MKL_INT yhint;                      // additional info about function
    MKL_INT nscoeff;                    // number of spline coefficients
    MKL_INT scoeffhint;                 // additional info about spline
                                        // coefficients
    MKL_INT bc_type;                    // boundary conditions type
    MKL_INT nbc;                        // number of boundary conditions
    MKL_INT ic_type;                    // internal conditions type
    MKL_INT nlim;                       // number of pairs of integration limits
    MKL_INT llimhint;                   // additional info about the structure
                                        // of left integration limits
    MKL_INT rlimhint;                   // additional info about the structure
                                        // of right integration limits
    MKL_INT rhint;                      // integration results storage format
    double x[2];                        // limits of the interpolation interval
    double y[NY*N];                     // function values
    double *ic;                         // internal conditions
    double *bc;                         // boundary conditions
    double scoeff[NSCOEFF];             // array of spline coefficients
    double llim[NLIM];                  // left  integration limits
    double rlim[NLIM];                  // right integration limits
    double *ldatahint, *rdatahint;      // additional info about the structure
                                        // of x and integration limits
    double r[NY*NLIM];                  // integration results
    double ref_r[NY*NLIM];              // reference integration results

    double left = LLIM_X, right = RLIM_X;
    double freq;
    double xx[N];
    double left_val[N-1], right_val[N-1];
    double left_der1[N-1], right_der1[N-1];
    double left_der2[N-1], right_der2[N-1];

    int i, j, errcode = 0;
    int errnums = 0;

    /***** Initializing parameters for Data Fitting task *****/

    sorder     = DF_PP_CUBIC;
    stype      = DF_PP_NATURAL;

    /***** Parameters describing interpolation interval *****/
    nx         = N;
    xhint      = DF_UNIFORM_PARTITION;
    /* Limits of interpolation interval are provided in case of uniform grid */
    x[0]       = left;
    x[1]       = right;

    /***** Parameters describing function *****/
    ny         = NY;
    yhint      = DF_NO_HINT;

    /***** Parameters describing spline coefficients storage *****/
    nscoeff    = NSCOEFF;
    scoeffhint = DF_NO_HINT;

    /***** Parameters describing boundary conditions type *****/
    bc_type    = DF_BC_FREE_END;
    /* No additional parameters are provided for Free-End boundary conditions */
    bc         = 0;

    /***** Parameters describing internal conditions type *****/
    /* No internal conditions are provided for natural cubic spline */
    ic_type    = DF_NO_IC;
    ic         = 0;

    /***** Parameters decsribing integration limits *****/
    nlim = NLIM;
    llimhint = DF_NON_UNIFORM_PARTITION;
    rlimhint = DF_NON_UNIFORM_PARTITION;

    /***** Additional information about the structure of integration
           limits *****/
    /* No additional info is provided */
    ldatahint = 0;
    rdatahint = 0;

    /***** Parameter dascribing integration results storage format *****/
    rhint = DF_NO_HINT;

    /***** Generate function y = sin(2 * Pi * freq * x) *****/
    errcode = dSinDataUniformGrid( y, FREQ, left, right, nx );
    CheckDfError(errcode);

    /***** Generate limits of integration intervals *****/
    errcode = dUniformRandSortedData( llim, LEFT_LLIM, RIGHT_LLIM, nlim );
    CheckDfError(errcode);
    errcode = dUniformRandSortedData( rlim, LEFT_RLIM, RIGHT_RLIM, nlim );
    CheckDfError(errcode);

    /***** Create Data Fitting task *****/
    errcode = dfdNewTask1D( &task, nx, x, xhint, ny, y, yhint );
    CheckDfError(errcode);

    /***** Edit task parameters for natural cubic spline construction *****/
    errcode = dfdEditPPSpline1D( task, sorder, stype, bc_type, bc,
                                 ic_type, ic, scoeff, scoeffhint );
    CheckDfError(errcode);

    /***** Construct natural cubic spline using STD method *****/
    errcode =  dfdConstruct1D( task, DF_PP_SPLINE, DF_METHOD_STD );
    CheckDfError(errcode);

    /***** Compute integrals *****/
    errcode = dfdIntegrate1D( task, DF_METHOD_PP, nlim, llim, llimhint,
                              rlim, rlimhint, ldatahint, rdatahint,
                              r, rhint );
    CheckDfError(errcode);

    /***** Check computed coefficients *****/
    errcode = dUniformData( xx, left, right, nx );
    CheckDfError(errcode);

    /***** Check spline values in break points *****/
    errcode = dCheckCubBreakPoints( nx, xx, ny, y, scoeff,
                                    left_val, right_val );
    if ( errcode < 0 ) errnums++;

    /***** Check that spline 1st derivatives are equal for left
           and right piece of the spline for each break point *****/
    errcode = dCheckCub1stDerConsistency( nx, xx, ny, scoeff,
                                          left_der1, right_der1 );
    if ( errcode < 0 ) errnums++;

    /***** Check that spline 2nd derivatives are equal for left
           and right piece of the spline for each break point *****/
    errcode = dCheckCub2ndDerConsistency( nx, xx, ny, scoeff,
                                          left_der2, right_der2 );
    if ( errcode < 0 ) errnums++;

    /***** Check boundary conditions *****/
    errcode = dCheckCubBC( nx, xx, ny, scoeff, bc_type, bc );
    if ( errcode < 0 ) errnums++;

    /***** Check integration results *****/
    errcode = dCheckCubIntegrRes( nx, xx, ny, scoeff, nlim, llim, rlim, r, ref_r );
    if ( errcode < 0 ) errnums++;

    /***** Print results *****/
    printf("Number of break points : %d\n", (int)nx);

    /***** Print given function *****/
    printf("\n  X           Y\n");

    for( j = 0; j < nx; j++ )
    {
        printf(" %+lf   %+lf\n", xx[j], y[j]);
    }

    /***** Print computed spline coefficients *****/
    printf("\nCoefficients are calculated for a polynomial of the form:\n\n");
    printf("Pi(x) = Ai + Bi*(x - x(i)) + Ci*(x - x(i))^2 + Di*(x - x(i))^3\n");
    printf("    where x(i) <= x < x(i+1)\n");
    printf("\nSpline coefficients for Y:\n");
    printf(" i    Ai            Bi            Ci            Di        ");
    printf("    P(X[i])       P(X[i+1]) ");
    printf("    P'(X[i])      P'(X[i+1])    P\"(X[i])      P\"(X[i+1])\n");

    for( j = 0; j < nx-1; j++ )
    {
        printf(" %1d %+11.6f   %+11.6f   %+11.6f   %+11.6f   %+11.6f   %+11.6f",
                j, scoeff[j*sorder], scoeff[j*sorder + 1],
                scoeff[j*sorder + 2], scoeff[j*sorder + 3],
                right_val[j], left_val[j]);
        printf("   %+11.6f   %+11.6f   %+11.6f   %+11.6f\n",
                right_der1[j], left_der1[j], right_der2[j], left_der2[j]);
    }

    printf("\nIntegration results for Y:\n");
    printf("Integration interval      Result\n");
    for ( j = 0; j < nlim; j++ )
    {
        printf(" ( %+4.1lf, %+4.1lf )        %+11.6lf\n", llim[j], rlim[j],
            r[j]);
    }

    /***** Delete Data Fitting task *****/
    errcode = dfDeleteTask( &task );
    CheckDfError(errcode);

    /***** Print summary of the test *****/
    if (errnums != 0)
    {
        printf("\n\nError: Computed natural cubic spline coefficients");
        printf(" or integrals are incorrect\n");
        return 1;
    }
    else
    {
        printf("\n\nComputed natural cubic spline coefficients");
        printf(" and integrals are correct\n");
    }

    return 0;
}

File: Examples/intel_examples/examples_core_f/datafittingf/source/dfdnaturalcubicspline.f
!===============================================================================
! Copyright 2010-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

!  Content:
!    Construction of natural cubic spline and spline-based integration
!  Example Program Text
!*******************************************************************************

      include 'mkl_df.f90'
      include "errcheck.inc"
      include "generatedata.inc"
      include "rescheck.inc"

      PROGRAM MKL_DF_TEST

      USE MKL_DF_TYPE
      USE MKL_DF
      USE DF_GENERATE_DATA
      USE DF_EXAMPLE_RESCHECK

      ! number of break points
      INTEGER,PARAMETER :: N          = 6
      ! number of functions
      INTEGER,PARAMETER :: NNY        = 1
      ! number of pairs of integration limits
      INTEGER,PARAMETER :: NNLIM      = 4
      ! total number of spline coefficients
      INTEGER,PARAMETER :: NNSCOEFF   = (NNY*(N-1)*DF_PP_CUBIC)
      ! left  limit of interpolation interval
      REAL(8),PARAMETER :: LLIM_X     =  0.0d0
      ! right limit of interpolation interval
      REAL(8),PARAMETER :: RLIM_X     =  7.0d0
      REAL(8),PARAMETER :: LEFT_LLIM  =  1.0d0
      REAL(8),PARAMETER :: RIGHT_LLIM =  3.0d0
      REAL(8),PARAMETER :: LEFT_RLIM  =  4.0d0
      REAL(8),PARAMETER :: RIGHT_RLIM =  5.5d0
      REAL(8),PARAMETER :: FFREQ      =  0.15d0

      ! Data Fitting task descriptor
      TYPE (DF_TASK) task
      ! spline order
      INTEGER :: sorder
      ! spline type
      INTEGER :: stype
      ! number of break points
      INTEGER :: nx
      ! additional info about break points
      INTEGER :: xhint
      ! number of functions
      INTEGER :: ny
      ! additional info about function
      INTEGER :: yhint
      ! spline coefficients storage format
      INTEGER :: scoeffhint
      ! boundary conditions type
      INTEGER :: bc_type
      ! internal conditions type
      INTEGER :: ic_type
      ! number of pairs of integration limits
      INTEGER :: nlim
      ! additional info about the structure of left integration limits
      INTEGER :: llimhint
      ! additional info about the structure of right integration limits
      INTEGER :: rlimhint
      ! integration results storage format
      INTEGER :: rhint
      ! limits of the interpolation interval
      REAL(8) :: x(2)
      ! break points in full format
      REAL(8) :: xx(N)
      ! function values
      REAL(8) :: y(NNY*N)
      ! array of spline coefficients
      REAL(8) :: scoeff(NNSCOEFF)
      ! left integration limits
      REAL(8) :: llim(NNLIM)
      ! right integration limits
      REAL(8) :: rlim(NNLIM)
      ! integration results
      REAL(8) :: r(NNY*NNLIM)
      ! reference integration results
      REAL(8) :: ref_r(NNY*NNLIM)

      REAL(8) :: left,right
      REAL(8) :: l_llim, r_llim, l_rlim, r_rlim
      REAL(8) :: freq
      REAL(8) :: left_val(N-1), right_val(N-1)
      REAL(8) :: left_der1(N-1), right_der1(N-1)
      REAL(8) :: left_der2(N-1), right_der2(N-1)

      INTEGER :: i,j,errnums
      INTEGER(4) :: errcode

      errcode = DF_STATUS_OK
      errnums = 0

      !***** Initializing parameters for Data Fitting task *****
      sorder = DF_PP_CUBIC
      stype  = DF_PP_NATURAL

      !***** Parameters describing interpolation interval *****
      nx = N
      xhint = DF_UNIFORM_PARTITION
      ! Limits of interpolation interval are provided in case of uniform grid
      left = LLIM_X
      right = RLIM_X
      x(1) = left
      x(2) = right

      !***** Parameters describing function *****
      ny = NNY
      yhint = DF_NO_HINT

      !***** Parameters describing spline coefficients storage *****
      scoeffhint = DF_NO_HINT

      !***** Parameters describing boundary conditions type *****
      bc_type = DF_BC_FREE_END

      !***** Parameters describing internal conditions type *****
      !***** No internal conditions are provided for natural cubic spline *****
      ic_type = DF_NO_IC

      !***** Parameters decsribing integration limits *****
      nlim = NNLIM
      llimhint = DF_NON_UNIFORM_PARTITION
      rlimhint = DF_NON_UNIFORM_PARTITION

      !***** Parameter dascribing integration results storage format *****
      rhint = DF_NO_HINT

      freq = FFREQ
      errcode = dSinDataUniformGrid( y, freq, left, freq, nx )
      CALL CheckDfError(errcode)

      !***** Generate limits of integration intervals *****
      l_llim = LEFT_LLIM
      r_llim = RIGHT_LLIM
      errcode = dUniformRandSortedData( llim, l_llim, r_llim, nlim )
      CALL CheckDfError(errcode)

      l_rlim = LEFT_RLIM
      r_rlim = RIGHT_RLIM
      errcode = dUniformRandSortedData( rlim, l_rlim, r_rlim, nlim )
      CALL CheckDfError(errcode)

      !***** Create Data Fitting task *****
      errcode = dfdnewtask1d( task, nx, x, xhint, ny, y, yhint )
      CALL CheckDfError(errcode)

      !***** Edit task parameters for natural cubic spline construction *****
      errcode = dfdeditppspline1d( task, sorder, stype, bc_type,         &
     &     ic_type=ic_type, scoeff=scoeff, scoeffhint=scoeffhint )
      CALL CheckDfError(errcode)

      !***** Construct natural cubic spline using STD method *****
      errcode = dfdconstruct1d( task, DF_PP_SPLINE, DF_METHOD_STD )
      CALL CheckDfError(errcode)

      !***** Compute integrals *****
      errcode = dfdintegrate1d( task, DF_METHOD_PP, nlim, llim,          &
     &    llimhint, rlim, rlimhint, r=r, rhint=rhint )
      CALL CheckDfError(errcode)

      !***** Check computed coefficients *****
      errcode = dUniformData( xx, left, right, nx );
      CALL CheckDfError(errcode)

      !***** Check spline values in break points *****
      errcode = dCheckCubBreakPoints( nx, xx, ny, y, scoeff,             &
     &      left_val, right_val )
      CALL CheckDfError(errcode)

      !***** Check that spline 1st derivatives are equal for left
      !      and right piece of the spline for each break point *****
      errcode = dCheckCub1stDerConsistency( nx, xx, ny, scoeff,          &
     &      left_der1, right_der1 )
      CALL CheckDfError(errcode)

      !***** Check that spline 2nd derivatives are equal for left
      !     and right piece of the spline for each break point *****/
      errcode = dCheckCub2ndDerConsistency( nx, xx, ny, scoeff,          &
     &      left_der2, right_der2 )
      CALL CheckDfError(errcode)

      !***** Check boundary conditions *****
      errcode = dCheckCubBC( nx, xx, ny, scoeff, bc_type )
      CALL CheckDfError(errcode)

      !***** Check results of integration *****
      errcode = dCheckCubIntegrRes( nx, xx, ny, scoeff,                  &
     & nlim, llim, rlim, r, ref_r )
      IF (errcode < 0) errnums = errnums+1

      !***** Print results *****
      WRITE (*,901) "Number of break points : ",nx

      !***** Print given function *****
      WRITE (*,902) " i    x(i)          y(i)"
      DO j = 1, nx
        WRITE (*,907) " ",j," ",xx(j),"   ",y(j)
      END DO

      !***** Print computed spline coefficients *****
      WRITE (*,904) "Coefficients are calculated for a polynomial of the &
     & form:",""
      WRITE (*,902) "Pi(x) = Ai + Bi*(x - x(i)) + Ci*(x - x(i))^2 + Di*(x&
     & - x(i))^3"
      WRITE (*,904) "    where x(i) <= x < x(i+1)",""
      WRITE (*,902) "Spline coefficients for Y:"
      WRITE (*,902) " i    Ai          Bi          Ci          Di        &
     &   P(x(i))     P(x(i+1))   P'(x(i))    P'(x(i+1))  P''(x(i))    P''&
     &(x(i+1))"

      DO j = 1, nx-1
        WRITE (*,910) " ", j, " ", scoeff(sorder*(j-1) + 1), " ",        &
     &      scoeff(sorder*(j-1) + 2), " ", scoeff(sorder*(j-1) + 3)," ", &
     &      scoeff(sorder*(j-1) + 4)
        WRITE (*,911) " ",right_val(j)," ",left_val(j)," ",              &
     &      right_der1(j)," ",left_der1(j)
        WRITE (*,912) " ",right_der2(j)," ",left_der2(j)
      END DO

      WRITE (*,904) "","Integration results for Y:"
      WRITE (*,902) " Integration interval               Result"

      DO j = 1, nlim
        WRITE (*,913) " ( ",llim(j)," , ",rlim(j)," )    ",r(j)
      END DO

      !***** Delete Data Fitting task *****
      errcode = dfdeletetask( task )
      CALL CheckDfError(errcode)

      !***** Print summary of the test *****
      IF (errnums /= 0) THEN
        WRITE (*,904) "","Error: Computed natural cubic spline           &
     &    coefficients or integrals are incorrect"
        STOP 1
      ELSE
        WRITE (*,904) "","Computed natural cubic spline coefficients and &
     &  integrals are correct"
      END IF
      STOP 0

901   FORMAT (A,I0)
902   FORMAT (/A)
904   FORMAT (/A/A)
907   FORMAT (A,I1,A,SP,F11.6,A,SP,F11.6)
910   FORMAT (A,I1,A,SP,F11.6,A,SP,F11.6,A,SP,F11.6,A,SP,F11.6,$)
911   FORMAT (A,SP,F11.6,A,SP,F11.6,A,SP,F11.6,A,SP,F11.6,$)
912   FORMAT (A,SP,F11.6,A,SP,F11.6)
913   FORMAT (A,SP,F11.6,A,SP,F11.6,A,SP,F11.6)

      END PROGRAM
